### DISCUSSION TITLE
How to change vector table address for my application?

### DISCUSSION THREAD

**Discussion by mucahitturk** [stm32duino/Arduino_Core_STM32#1991] [General]
Hello I developed custom bootloader application and test application in keil ide and it is working for stm32f103c8t6 .

There is a project that developed with arduino ide and i just want to change start position of this program.

For test application, i added VECT_TAB_OFFSET 0x00002000 and i changed the rom1 start address from 0x8000000 to 0x8002000 in keil ide.In this configuration test application is working when i uploaded using bootloader app.

How can i set same configuration in the arduino ide?

**Comment by fpistm** [comment]
Hi @mucahitturk 
which core you used as it seems you already open an issue to other core:
https://github.com/stm32duino/Arduino_Core_STM32/discussions/1991

**Reply by mucahitturk** [reply]
Actually, i do not know the core. I just download from board manager in arduino ide. How can I learn which I download core?

**Reply by fpistm** [reply]
Which link you used?

**Reply by mucahitturk** [reply]
http://dan.drown.org/stm32duino/package_STM32duino_index.json

**Reply by fpistm** [reply]
OK. So this is not this core. Refers to this issue. Hope you will get answer else you can try the forum in the section libMaple:
https://www.stm32duino.com/viewforum.php?f=63

**Reply by mucahitturk** [reply]
Hello,
I changed the link. New link is https://github.com/stm32duino/BoardManagerFiles/raw/main/package_stmicroelectronics_index.json.
How can I change vector table address of my application now?

**Reply by fpistm** [reply]
Which target you used in the menu ?

**Comment by mucahitturk** [comment]
![Adsız](https://user-images.githubusercontent.com/60607054/229729362-cc0aad6f-193f-44aa-ad3a-1873be4c6f3d.png)

**Reply by fpistm** [reply]
You can use `boards.local.txt` feature:
https://arduino.github.io/arduino-cli/0.31/platform-specification/#boardslocaltxt

For your case, it should contains this:
```
GenF1.menu.pnum.GENERIC_F103C8TX.build.flash_offset=0x2000
GenF1.menu.pnum.GENERIC_F103C8TX.build.bootloader_flags=-DVECT_TAB_OFFSET={build.flash_offset}
```

Then in your build log you should see this in the log:
```
-DVECT_TAB_OFFSET=0x2000 (multiple time)
--defsym=LD_FLASH_OFFSET=0x2000
```

Then you will have to flash manually at the correct address.

I'm currently working to manage upload offset using the STM32CubeProgrammer option (#1993), so upload will be manage even with an offset.
And no more need to specify `VECT_TAB_OFFSET`, only the `flash_offset` will be required.

**Reply by mucahitturk** [reply]
I added the two commands to board.txt in 2.5.0 folder. After added, i closed then opened arduino ide. I compiled and uploaded to board and the log screen is in the under.

![asd](https://user-images.githubusercontent.com/60607054/229764317-ca0dae86-2b3b-424a-905b-c1531b7a12da.PNG)

I saw in the log what you said after added commands to boards.txt. I uploaded program using bootloader and it doesn't works with these changes. Board.txt is in the under.

![asd2](https://user-images.githubusercontent.com/60607054/229766676-c6316a85-6833-4614-8bc0-ef446f259ea6.PNG)

Have I applied these changes correctly?

**Reply by fpistm** [reply]
As stated before, You have to upload manually at the specific address.

**Reply by mucahitturk** [reply]
I upload using bootloader application and ST-LINK Utility program.Still it is not working. 
In the first picture, there is a bootloader application code that start at adress 0x0800 0000.

![bootloader app](https://user-images.githubusercontent.com/60607054/229774312-62ad998a-221c-4102-bc00-e126d2d15f0b.PNG)

In the second picture, there is an application code that start at address 0x0800 2000.
![normal app](https://user-images.githubusercontent.com/60607054/229774329-27718e6c-7aa7-4a61-9149-5a2c08f95b18.PNG)

**Reply by fpistm** [reply]
well, when doing this it works on my side. It start the BL, then it starts application or if I upload within STM32CubeProgrammer and said to start at offset it works.

**Reply by mucahitturk** [reply]
In my side, it doesn't work. Why is this happening? I just blink the led on the board that is connected to pin PC13 on the test application.
I develop one project using keil ide and it is working same like i developed app in arduino ide.

**Reply by fpistm** [reply]
> In my side, it doesn't work. Why is this happening?

If I know I will tell you. But as said don't know.
With 2.5.0 core version, defining the build.flash_offset allows to have the flash configuration in the ldscipt.ld properly configured. Then thanks the VECT_TAB_OFFSET, define the VTOR at the right address that's all.
Maybe a STLink utility issue. Try with the STM32cubeProgrammer and with an updated STLink firmware.

**Reply by mucahitturk** [reply]
Is there a any .bin file to test with address at 0x8002000.

**Reply by fpistm** [reply]
what is your board exactly? LED pin?

**Reply by mucahitturk** [reply]
The board is STM32f103c8t6 - Blue Pill. Led is connected to GPIOC13.

**Reply by fpistm** [reply]
If it is a BluePill then simply select BluePill F103C8 in the menu and select the `Maple DFU Bootloader 2.0` upload method, this will set the offset to 0x2000. then flash manually the binary at offset.

**Reply by fpistm** [reply]
[sketch_apr4a.ino.bin.zip](https://github.com/stm32duino/Arduino_Core_STM32/files/11149245/sketch_apr4a.ino.bin.zip)

**Reply by fpistm** [reply]
Just for my information what is the sketch you build?
